File: ge-develop/examples/es/dynamic_input/cpp/src/make_concatv2_graph.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of 
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, 
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include "es_showcase.h"// es构图方式
#include "es_ConcatV2.h"
#include "utils.h"
#include <memory>
#include "ge/ge_api.h"
using namespace ge;
using namespace ge::es;
namespace {
es::EsTensorHolder MakeConcatV2Graph(es::EsTensorHolder tensor1, es::EsTensorHolder tensor2, es::EsTensorHolder tensor3,
                                         es::EsTensorHolder concat_dim) {
  return ConcatV2({tensor1, tensor2, tensor3}, concat_dim, 3);
}
}
namespace es_showcase {
int RunGraph(ge::Graph &graph, const std::vector<ge::Tensor> &inputs,
             const std::string &output_prefix) {
  ge::Utils::PrintTensorsToFile(inputs, "input");
  std::map<ge::AscendString, ge::AscendString> options;
  auto *s = new (std::nothrow) ge::Session(options);
  if (s == nullptr) {
    std::cout << "Global session not ready" << std::endl;
    return -1;
  }
  static uint32_t next =0;
  const uint32_t graph_id = next++;
  auto ret = s->AddGraph(graph_id, graph);
  if (ret != ge::SUCCESS) {
    std::cout << "AddGraph failed" << std::endl;
    delete s;
    return -1;
  }
  std::vector<ge::Tensor> outputs;
  ret = s->RunGraph(graph_id, inputs, outputs);
  if (ret != ge::SUCCESS) {
    std::cout << "RunGraph failed" << std::endl;
    (void)s->RemoveGraph(graph_id);
    delete s;
    return -1;
  }
  (void)s->RemoveGraph(graph_id);
  ge::Utils::PrintTensorsToFile(outputs, output_prefix);
  delete s;
  return 0;
}
void MakeConcatV2GraphByEsAndDump() {
  std::unique_ptr<ge::Graph> graph = MakeConcatV2GraphByEs();
  graph->DumpToFile(ge::Graph::DumpFormat::kOnnx, ge::AscendString("make_concatv2_graph"));
}
std::unique_ptr<ge::Graph> MakeConcatV2GraphByEs() {
  // 1、创建图构建器
  auto graph_builder = std::make_unique<EsGraphBuilder>("MakeConcatV2Graph");
  // 2、创建输入节点 在第0维拼接tensor1、tensor2、tensor3
  auto tensor1 = graph_builder->CreateInput(0, "tensor1", ge::DT_FLOAT, ge::FORMAT_ND, {8, 64, 128});
  auto tensor2 = graph_builder->CreateInput(1, "tensor2", ge::DT_FLOAT, ge::FORMAT_ND, {2, 64, 128});
  auto tensor3 = graph_builder->CreateInput(2, "tensor3", ge::DT_FLOAT, ge::FORMAT_ND, {6, 64, 128});
  auto concat_dim = graph_builder->CreateScalar(0);
  /*
    ConcatV2原型注释：
    REG_OP(ConcatV2)
      .DYNAMIC_INPUT(x, TensorType({BasicType(), DT_BOOL, DT_STRING}))
      .INPUT(concat_dim, TensorType::IndexNumberType())
      .OUTPUT(y, TensorType({BasicType(), DT_BOOL, DT_STRING}))
      .ATTR(N, Int, 1)
      .OP_END_FACTORY_REG(ConcatV2)
  */
  auto result = MakeConcatV2Graph(tensor1, tensor2, tensor3, concat_dim);
  // 3、设置输出
  (void) graph_builder->SetOutput(result, 0);
  // 4、构建图
  auto graph = graph_builder->BuildAndReset();
  return graph;
}
int MakeConcatV2GraphByEsAndRun() {
  std::unique_ptr<ge::Graph> graph = MakeConcatV2GraphByEs();
  std::vector<ge::Tensor> inputs;
  std::vector<float> tensor1_data(8 * 64 * 128, 1.0f);
  std::vector<float> tensor2_data(2 * 64 * 128, 2.0f);
  std::vector<float> tensor3_data(6 * 64 * 128, 3.0f);
  inputs.push_back(*ge::Utils::StubTensor<float>(tensor1_data, {8, 64, 128}));
  inputs.push_back(*ge::Utils::StubTensor<float>(tensor2_data, {2, 64, 128}));
  inputs.push_back(*ge::Utils::StubTensor<float>(tensor3_data, {6, 64, 128}));
  return RunGraph(*graph, inputs, "ConcatV2");
}
}

File: ge-develop/examples/es/dynamic_input/python/src/make_concatv2_graph.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
# ----------------------------------------------------------------------------
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of 
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, 
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------

import numpy as np

from ge.es.graph_builder import GraphBuilder, TensorHolder
from ge.graph import Tensor
from ge.graph.types import DataType, Format
from ge.graph import Graph, DumpFormat
from ge.ge_global import GeApi
from ge.session import Session
from ge.es.all import ConcatV2


def build_concat_graph():
    # 1、创建图构建器
    builder = GraphBuilder("MakeConcatV2Graph")
    # 2、创建图输入节点
    tensor_holder1 = builder.create_input(
        index=0,
        name="tensor1",
        data_type=DataType.DT_FLOAT,
        shape=[8, 64, 128]
    )
    tensor_holder2 = builder.create_input(
        index=1,
        name="tensor2",
        data_type=DataType.DT_FLOAT,
        shape=[2, 64, 128]
    )
    tensor_holder3 = builder.create_input(
        index=2,
        name="tensor3",
        data_type=DataType.DT_FLOAT,
        shape=[6, 64, 128]
    )
    tensor_holder_list = [tensor_holder1, tensor_holder2, tensor_holder3]
    concat_tensor_holder = ConcatV2(tensor_holder_list, 0, N=3)
    # 3、设置图输出节点
    builder.set_graph_output(concat_tensor_holder, 0)
    # 4、构建图
    return builder.build_and_reset()


def dump_concat_graph(graph):
    graph.dump_to_file(format=DumpFormat.kOnnx, suffix="make_concat_graph")


def run_graph(graph) -> int:
    config = {
        "ge.exec.deviceId": "0",
        "ge.graphRunMode": "0"  # 0: 图模式, 1: 单算子模式
    }
    ge_api = GeApi()
    ret = ge_api.ge_initialize(config)
    if ret != 0:
        print(f"GE初始化失败，返回码: {ret}")
        return ret
    print("GE环境初始化成功 (Device ID: 0)")

    try:
        # 2. 创建Session
        session = Session()
        # 3. 添加图到Session
        graph_id = 1
        ret = session.add_graph(graph_id, graph)
        if ret != 0:
            print(f"添加图失败，返回码: {ret}")
            return ret
        print(f"图已添加到Session (Graph ID: {graph_id})")

        # 4. 准备输入数据
        tensor1_data = np.full(8 * 64 * 128, 1.0, dtype=np.float32)
        tensor2_data = np.full(2 * 64 * 128, 2.0, dtype=np.float32)
        tensor3_data = np.full(6 * 64 * 128, 3.0, dtype=np.float32)
        tensor1 = Tensor(
            tensor1_data.flatten().tolist(),
            None,
            DataType.DT_FLOAT,
            Format.FORMAT_ND,
            shape=[8, 64, 128]
        )
        tensor2 = Tensor(
            tensor2_data.flatten().tolist(),
            None,
            DataType.DT_FLOAT,
            Format.FORMAT_ND,
            shape=[2, 64, 128]
        )
        tensor3 = Tensor(
            tensor3_data.flatten().tolist(),
            None,
            DataType.DT_FLOAT,
            Format.FORMAT_ND,
            shape=[6, 64, 128]
        )
        input_tensor = [tensor1, tensor2, tensor3]
        # 5. 运行图
        ret = session.run_graph(graph_id, input_tensor)
        print("[Info] 图运行成功！")
        for idx, tensor in enumerate(ret, start=1):
            print(f"Tensor{idx}详情： {tensor}")
        return 0

    except Exception as e:
        print(f"[Error] 执行过程中出错: {e}")
        import traceback
        traceback.print_exc()
        return -1
    finally:
        # 6. 清理GE环境
        print("[Info] 清理GE环境...")
        ge_api.ge_finalize()
        print("[Success] GE环境已清理")


graph = build_concat_graph()
dump_concat_graph(graph)
run_graph(graph)
